### DISCUSSION TITLE
Code Hike v1.0

### DISCUSSION THREAD

**Discussion by pomber** [code-hike/codehike#413] [Ideas and feedback]
I'm working on the next major version of Code Hike. The plan is to release it in April 2024 with a new website, playground, and docs. Beta versions will be available before that.

A key focus for v1.0 is more flexibility. We're introducing features like custom annotations and layouts. This will also make Code Hike smaller since we can move some of the current features to copypastable components that you can adapt to your needs, like what https://ui.shadcn.com/ does.

**I'm going to list my ideas for v1.0 in this discussion but I'd love to hear yours too. Your early feedback is crucial to influence v1.0's development before we get too far along.**

The new website and docs (work in progress): https://v1.codehike.org/

**Comment by pomber** [comment]
## Custom annotations

Use your own components (even server components) as annotations.

For example:

````mdx
import { MyComponent } from "./MyComponent"

```js
console.log(1)
// !MyComponent foo
console.log(2)
```

````

`MyComponent` will receive the syntax highlighted `console.log(2)` as `children` and "foo" as another prop.

---

Here's an example of a tooltip annotation

https://github.com/code-hike/codehike/assets/1911623/ec94d7cc-8251-48f4-913a-b28830a55146

For this kind of annotation, I still trying to figure out a good syntax to pass extra markdown as a prop to the component (the content of the tooltip). 

---

Another example that uses React Context:

https://github.com/code-hike/codehike/assets/1911623/f8e3f122-74a5-4efd-9cde-f65900111bd0

**Comment by pomber** [comment]
## Custom Layouts

Instead of providing components like `<CH.Scrollycoding>`, `<CH.Spotlight>`, and `<CH.Slideshow>` provide one component to rule them all: `<Hike>`.

`Hike` will have a prop where you pass the component you want to use as the layout. 

Considerations:
- Provide the current layouts as copypastable components (like https://ui.shadcn.com/)
- Provide utils to manage the steps and reduce the amount of work those copypastable should do


For example, a custom layout with a UI similar to Instagram Stories:

https://github.com/code-hike/codehike/assets/1911623/0d0f9e59-4517-430e-b1dd-b86e65581810

Live demo: https://ch-stories.vercel.app/

**Comment by pomber** [comment]
## Code transitions

New transition between two pieces of code that works better with small changes. 


https://github.com/code-hike/codehike/assets/1911623/38a3595e-c210-4940-9f2c-d0eac430a1ab

**Reply by tomByrer** [reply]
Other art & markup ideas:
https://antfu.me/posts/shiki-magic-move

**Comment by DanielFGray** [comment]
absolutely love this! especially the last one with more fine-grained animations

[my blog](https://danielfgray.com) makes heavy use of code hike, and if you are offering early access I would love to be considered for it

**Reply by pomber** [reply]
will start adding more info here https://v1.codehike.org/

**Comment by eric-horodyski** [comment]
Code folding, please!

**Reply by pomber** [reply]
Coming soon (as a custom annotation, so you can customize it as much as you want).

I already have this one working:

https://github.com/code-hike/codehike/assets/1911623/e9d2106f-f30f-483c-b911-edd4d3020fe9

**Reply by eric-horodyski** [reply]
Awesome! As far as customization goes, can we set the initial state (either collapsed or expanded)? I haven't had much time to read into what custom annotations would look like/how they're constructed.

**Reply by pomber** [reply]
Yes, you can set the initial state. I'm still refining the API, but as soon I have something stable I'll start adding examples

Edit: https://v1.codehike.org/docs/code/collapse

**Comment by Jokcy** [comment]
Seems there is no branch for this new release right now?

**Reply by pomber** [reply]
work is happening here https://github.com/code-hike/v1

**Reply by stern-shawn** [reply]
If we'd like to make contributions, would it make the most sense to do so against the new repo, or this one? 😄

**Reply by pomber** [reply]
the new one, feel free to open any issue with more details there

**Comment by chrisjaure** [comment]
Awesome to see the next iteration of Code Hike! Thanks for sharing 🙂 

One thing I would love to see from a tool like this is to be able to make relationships more discoverable. I've considered using monaco editor for this in the past. These are features I miss the most from VSCode that would be cool to see in Code Hike:
- [Peek](https://code.visualstudio.com/docs/editor/editingevolved#_peek) -- show related code inline
- [Go to Definition](https://code.visualstudio.com/docs/editor/editingevolved#_go-to-definition) -- quickly jump through tabs
- [Hover Information](https://code.visualstudio.com/docs/languages/typescript#_hover-information) -- for Typescript, get type information that may not be present in the visible code

I realize the changes you're making now would support some of these. At the least they would make some cool copy/paste examples.

Thanks again for sharing Code Hike with us!

**Reply by pomber** [reply]
I think you will be able to do all three. 

I'll add a copypastable example for the typescript one ([using twoslash](https://www.typescriptlang.org/dev/twoslash/)) soon. But it will probably only work with React Server Components.

Edit: https://v1.codehike.org/docs/code/twoslash

**Comment by ethan-sh-lee** [comment]
Hi pomber! thanks for your work :)

I would like to request support for intergration with [velite](https://velite.js.org/)

Thanks again.

**Reply by pomber** [reply]
Hi! it's on my to-do list. 

Apparently, [velite doesn't support the `recmaPlugins` option](https://velite.js.org/reference/config#markdown) that Code Hike needs. If you want to help you can submit an issue on the velite repo. Here are similar issues I submitted to [nextra](https://github.com/shuding/nextra/issues/2855) and [docusaurus](https://github.com/code-hike/examples/pull/8).

**Comment by fivetran-jimmyhooker** [comment]
Would love for you to consider [markdoc](https://markdoc.dev/) support!

**Reply by pomber** [reply]
hey, I'm open to add markdoc support if anyone wants to sponsor the work
